metagraph/plugins/numpy/types.py

```diff
@@ -230,17 +230,42 @@ def __init__(self, data, *, mask=None, node_ids=None):
                 len(id2pos) == len(data), f"node_ids must be the same length as data"
             )
 
-    def __getitem__(self, node_id):
+    def __len__(self):
+        return self.num_nodes
+
+    def _get_single_item(self, node_id):
         if self.mask is not None:
-            if self.mask[node_id]:
+            if not self.mask[node_id]:
                 raise ValueError(f"node {node_id} is not in the NodeMap")
         elif self.id2pos is not None:
             if node_id not in self.id2pos:
                 raise ValueError(f"node {node_id} is not in the NodeMap")
-            pos = self.id2pos[node_id]
-            return self.value[pos]
+            node_id = self.id2pos[node_id]
         return self.value[node_id]
 
+    def _get_multiple_items(self, node_ids: np.ndarray):
+        if self.mask is not None:
+            if (node_ids >= len(self.mask)).any() or not self.mask[node_ids].all():
+                out_of_bounds_ids = node_ids[node_ids >= len(self.mask)]
+                mask_missing_ids = np.flatnonzero(~self.mask[node_ids])
```

Comment:
self.mask[node_ids] will fail here if there are any out_of_bounds_ids. It might be easier to first check for out of bounds and raise. Only after that check for missing ids.

Reply:
https://github.com/metagraph-dev/metagraph/pull/133/commits/48f1e38222f2d2c982798f2f06b241c509abcbdb implements this suggestion.